### ISSUE TITLE
Data Retention

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#4029]
**What is the goal / desired outcome?**

Be able to purge data, both for data sensitivity and storage cost reasons.

**If there is a way to accomplish this today via workaround, what does that require?**

Not really, at least it's very awkward.

**(Optional) What is your proposal for a feature to solve this?**

### API
To solve this we'll add a notion of a `RetentionSpec` which looks like this:

```
message RetentionSpec {
  // Duration indicates that commits should expire after the duration has
  // elapsed.
  google.protobuf.Duration duration = 1;
  // Transitive indicates that commits should expire when any of their
  // provenant commits does.
  bool transitive = 2;
}
```

Retention specs will be attached to repos via `CreateRepoRequest`. Commits that are created on a repo with a retention spec will have their `Expires` field set based on the details of the spec. Retention specs will also be attached to pipelines, which just passes the retention spec through to the output repo. I don't think we should add a way to set expirations on individual commits, I think it's likely to be confusing and lead to people retaining data they thought would be purged, and I can't think of a compelling use case for it.

### Implementation

When a commit expires we unreference the tree associated with it, which eventually leads to its content being GCed. This is similar to what we do in `DeleteCommit` except that it doesn't recursively delete subvenance, and the commit can still be accessed by `*Commit` requests, it can't be accessed by `*File` requests.

To make this work we'll need to add a pfs master, similar to the pps master, that watches the commits collection for expired commits (using an index on the `Expires` field) and unrefences their trees when the expiration passes. Expiration should block until the commit has been consumed by all of its downstream pipelines.

### Usage

Setting a retention spec alone is only going to have the effect the user wants for overwrite workloads where new commits delete old data in addition to adding new data. In an append workload, old commits will expire, but their content will already have been duplicated in later commits. However, setting retention means that when you call `delete file` and then wait for the expiration time to elapse, the files you deleted will be gone.

There are a couple of other scenarios in which users can inadvertently hold on to data they want to expire. The obvious one is pipelines which copy their input data, users will have to be careful to set such pipelines with a transitive retention spec so that data gets expired when its input does. The other potential issue is if people copy data, there's no particular support for this, users just need to be careful to add retention policies to copied data they don't want to retain.

I've layed out what I think the proto changes for this feature are going to look like in he retention branch: https://github.com/pachyderm/pachyderm/tree/retention

**Comment by jdoliner** [comment]
Adding onto this, to make the automated workload smoother we should also make it so that when committing new files, files in the parent commit that are older than the retention policy don't get checkpointed into new commits.

**Comment by RaananHadar** [comment]
This is a very important feature for large datasets. 

I do have a simple point though, it should be possible to manually set or update the retention policy for commits that have finished processing. Maybe even apply them to all the current commits in a repo retroactively?

**Comment by RaananHadar** [comment]
@jdoliner can you please consider this for 1.11? sounds like one of the more important features you guys thought of.

**Comment by jdoliner** [comment]
@RaananHadar we definitely see this as a very important feature of Pachyderm. It's also a very difficult feature, because so much of Pachyderms architecture is built for retaining every version of the data, forgetting data cuts counter to a lot of that architecture. The other thing that makes this tricky is that we're presently working on a rewrite of our storage system which means that if we implement this for the existing system we'll have to reimplement it in the new system which would be duplicating a lot of work. For that reason it makes more sense for us to implement this after the new storage layer lands. We're targeting that for either our next release or the one after, whenever it's ready. And then we'll start working on this immediately. Sorry I can't offer you a better timeline, but that's what makes the most sense for the overall roadmap of the project.

**Comment by RaananHadar** [comment]
@jdoliner thank you for the explanation. Makes sense.

**Comment by szeitlin** [comment]
Looking forward to having this! Not a rush for us though (yet).

**Comment by TimovNiedek** [comment]
Hi,

It would be great if this could just be configured in the pipeline specification. One way would be to specify the TTL in a duration, but for some use cases where the data timing is not yet known in advance it would be useful to be able to specify a number of commits to keep.

Another nice to have would be to still keep track of the provenance such that the deleted data can be reproduced if needed.

Is there already a known workaround? What would happen if we just set a TTL on the objects in the bucket itself - how would pachyderm behave? We're getting closer to this being a blocking issue for us, so any workarounds would be much appreciated.

Thanks for the discussion!

**Comment by jamesthegiantpeach** [comment]
Something that might be relevant to this issue:

Currently we’re dealing with a GDPR issue that seems incoherent to me. The combination of GDPR and standard IRBs for healthcare related studies makes it such that a patient needs to always be able to delete their data. That is, within something like a month after the patient has requested to be removed from a study, all of their data needs to be out of the system. They also require strong data provenance which is obviously competing with the data deletion requirements. 

This is probably more of an issue of incoherent regulations, but I thought I’d drop my two cents. Being able to delete data easily ends up being pretty important for dealing with GDPR.

**Comment by JoeyZwicker** [comment]
Yeah the law is definitely not clear on these things. I think the overall plan for this feature is that the data will be actually deleted, but the commit/lineage structure will still remain. So you'll have some amount of lineage info still yet the actual data will be gone. There is a lot of complexity here we're still trying to design and sort out so none of what I'm describing is "final" either.

Joey Zwicker
Founder, Pachyderm

**Comment by Obsecurus** [comment]
Has there been any further discussion/documentation on this? I'm using AWS S3 as the backing store and it would be nice to have a strategy for using intelligent tiering or migration to AWS Glacier.
